File: src/workflow.ts
import { readFileSync, existsSync } from 'node:fs'
import { join, resolve } from 'node:path'
import { load as yamlLoad } from 'js-yaml'
import type { WorkflowDefinition } from './models'

export function loadWorkflow(path: string | null): WorkflowDefinition {
  const resolvedPath = path ? resolve(path) : join(process.cwd(), 'WORKFLOW.md')

  if (!existsSync(resolvedPath)) {
    throw new Error(`Workflow file not found: ${resolvedPath}`)
  }

  const raw = readFileSync(resolvedPath, 'utf-8')
  const { config, promptBody } = splitFrontMatter(raw)

  const validatedConfig = config !== null ? config : {}
  if (validatedConfig !== null && (typeof validatedConfig !== 'object' || Array.isArray(validatedConfig))) {
    throw new Error('YAML front matter must decode to a map/object')
  }

  return {
    config: validatedConfig as Record<string, unknown>,
    promptTemplate: promptBody.trim(),
  }
}

function splitFrontMatter(raw: string): { config: unknown; promptBody: string } {
  if (!raw.startsWith('---')) {
    return { config: null, promptBody: raw }
  }

  const rest = raw.slice(3)
  const endIdx = rest.indexOf('\n---')
  if (endIdx === -1) {
    return { config: null, promptBody: raw }
  }

  const yamlText = rest.slice(0, endIdx)
  const body = rest.slice(endIdx + 4)

  const config = yamlLoad(yamlText)
  return { config, promptBody: body }
}

File: python/src/symphony/workflow.py
import os
from pathlib import Path
from typing import Any

import yaml

from symphony.exceptions import (
    MissingWorkflowFile, WorkflowParseError, WorkflowFrontMatterNotAMap,
)
from symphony.models import WorkflowDefinition


def load_workflow(path: str | Path | None) -> WorkflowDefinition:
    """Load and parse WORKFLOW.md.

    If path is None, default to ./WORKFLOW.md in cwd.
    """
    if path is None:
        path = Path.cwd() / "WORKFLOW.md"
    else:
        path = Path(path)

    if not path.is_file():
        raise MissingWorkflowFile(f"Workflow file not found: {path}")

    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as e:
        raise MissingWorkflowFile(f"Cannot read workflow file: {e}")

    config, prompt_body = _split_front_matter(raw)

    if config is not None:
        if not isinstance(config, dict):
            raise WorkflowFrontMatterNotAMap(
                "YAML front matter must decode to a map/object"
            )
    else:
        config = {}

    prompt_template = prompt_body.strip()

    return WorkflowDefinition(config=config, prompt_template=prompt_template)


def _split_front_matter(raw: str) -> tuple[dict[str, Any] | None, str]:
    """Split raw markdown into (front_matter_dict, body_string).

    Returns (None, body) if no front matter delimiter found.
    """
    if not raw.startswith("---"):
        return None, raw

    rest = raw[3:].lstrip("\n")
    end_idx = rest.find("\n---")
    if end_idx == -1:
        return None, raw

    yaml_text = rest[:end_idx]
    body = rest[end_idx + 4:]

    try:
        config = yaml.safe_load(yaml_text)
    except yaml.YAMLError as e:
        raise WorkflowParseError(f"Invalid YAML front matter: {e}")

    return config, body
